### PR TITLE
INT: Add an intention to import constants in match arms

### LINKED ISSUE #7815: Add an intention to import constants in match arms
## Environment

* **IntelliJ Rust plugin version:** 0.4.153.4056-212
* **Rust toolchain version:** 1.54.0 (a178d0322 2021-07-26) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2021.2 (CL-212.4746.93)
* **Operating system:** Linux 5.13.13-arch1-1
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description

Names in match arms are treated as variables unless there is a constant in scope with the same name. There should be an intention to import a constant of the name if such a constant exists.

## Steps to reproduce

```rust
mod x {
    const C: u32 = 1;
    
    mod y {
        fn f() {
            match 1u32 {
                C => { },  // add intention to import C here
                _ => { },
            }
        }
    }
}
```

### PR DESCRIPTION
**Pull request by dima74**
Fixes #7815

Consider match arm:
```rust
match ... {
    Foo => {}
    _ => {}
}
```

Here `Foo` can be constant or enum variant if any is in scope, otherwise it will be name of variable (irrefutable pattern). Therefore we don't show unresolved reference error for `Foo`, though user actually may want `Foo` to be constant but doesn't import it yet. This PR adds `Import` intention for such cases.

changelog: Add an intention to import constants in match arms

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt [hunk 1]
```diff
@@ -96,12 +112,8 @@ class RsUnresolvedReferenceInspection : RsLocalInspectionTool() {
...
 
         val highlightedElement = element.referenceNameElement ?: element
-        registerProblem(
-            highlightedElement,
-            description,
-            ProblemHighlightType.LIKE_UNKNOWN_SYMBOL,
-            *fixes.toTypedArray()
-        )
+        val highlightType = if (context?.type == AutoImportFix.Type.PAT_BINDING) INFORMATION else LIKE_UNKNOWN_SYMBOL
+        registerProblem(highlightedElement, description, highlightType, *fixes.toTypedArray())
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt:116]: This code contains following problems:
- you always register a problem with `INFORMATION` hightlighting type, even in batch mode. In leads to
```
com.intellij.diagnostic.PluginException: Tool #RsUnresolvedReference (class org.rust.ide.inspections.RsUnresolvedReferenceInspection) registers 'INFORMATION'-level problem in batch mode on FILE. Warnings of the 'INFORMATION' level are invisible in the editor and should not become visible in batch mode. Moreover, since 'INFORMATION'-level fixes act more like intention actions, they could e.g. change semantics and thus should not be suggested for batch transformations [Plugin: org.rust.lang]
	at com.intellij.codeInspection.InspectionEngine$2.registerProblem(InspectionEngine.java:261)
	at com.intellij.codeInspection.ProblemsHolder.registerProblem(ProblemsHolder.java:52)
	at org.rust.ide.inspections.RsProblemsHolder.registerProblem(RsLocalInspectionTool.kt:86)
	at org.rust.ide.inspections.RsUnresolvedReferenceInspection.registerProblem(RsUnresolvedReferenceInspection.kt:116)
	at org.rust.ide.inspections.RsUnresolvedReferenceInspection.access$registerProblem(RsUnresolvedReferenceInspection.kt:23)
	at org.rust.ide.inspections.RsUnresolvedReferenceInspection$buildVisitor$1.visitPatBinding(RsUnresolvedReferenceInspection.kt:78)
	at org.rust.lang.core.psi.impl.RsPatBindingImpl.accept(RsPatBindingImpl.java:21)
	at org.rust.lang.core.psi.impl.RsPatBindingImpl.accept(RsPatBindingImpl.java:26)
	at com.intellij.codeInspection.InspectionEngine.acceptElements(InspectionEngine.java:81)
	at com.intellij.codeInspection.InspectionEngine.createVisitorAndAcceptElements(InspectionEngine.java:56)
	at com.intellij.codeInspection.InspectionEngine.lambda$inspectElements$9(InspectionEngine.java:282)
	at com.intellij.concurrency.JobLauncherImpl.lambda$processImmediatelyIfTooFew$2(JobLauncherImpl.java:140)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:608)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:683)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:639)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:607)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.concurrency.JobLauncherImpl.lambda$processImmediatelyIfTooFew$3(JobLauncherImpl.java:136)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:915)
	at com.intellij.concurrency.JobLauncherImpl.processImmediatelyIfTooFew(JobLauncherImpl.java:147)
	at com.intellij.concurrency.JobLauncherImpl.invokeConcurrentlyUnderProgress(JobLauncherImpl.java:47)
	at com.intellij.concurrency.JobLauncher.invokeConcurrentlyUnderProgress(JobLauncher.java:51)
	at com.intellij.codeInspection.InspectionEngine.lambda$inspectElements$10(InspectionEngine.java:301)
	at com.intellij.codeInspection.InspectionEngine.withSession(InspectionEngine.java:226)
	at com.intellij.codeInspection.InspectionEngine.inspectElements(InspectionEngine.java:242)
	at com.intellij.codeInspection.InspectionEngine.inspectElements(InspectionEngine.java:215)
	at com.intellij.codeInspection.InspectionEngine.inspectEx(InspectionEngine.java:121)
	at com.intellij.codeInspection.ex.GlobalInspectionContextImpl.lambda$inspectFile$19(GlobalInspectionContextImpl.java:493)
	at com.intellij.codeInspection.ex.InspectionProfileWrapper.runWithCustomInspectionWrapper(InspectionProfileWrapper.java:89)
	at com.intellij.codeInspection.ex.GlobalInspectionContextImpl.inspectFile(GlobalInspectionContextImpl.java:490)
	at com.intellij.codeInspection.ex.GlobalInspectionContextImpl.lambda$runTools$6(GlobalInspectionContextImpl.java:353)
	at com.intellij.openapi.project.DumbService.lambda$runReadActionInSmartMode$0(DumbService.java:113)
	at com.intellij.openapi.project.DumbService.lambda$runReadActionInSmartMode$1(DumbService.java:157)
	at com.intellij.openapi.application.impl.ApplicationImpl.runReadAction(ApplicationImpl.java:941)
	at com.intellij.openapi.application.ReadAction.compute(ReadAction.java:68)
	at com.intellij.openapi.project.DumbService.runReadActionInSmartMode(DumbService.java:150)
	at com.intellij.openapi.project.DumbService.runReadActionInSmartMode(DumbService.java:113)
	at com.intellij.openapi.project.DumbService.tryRunReadActionInSmartMode(DumbService.java:130)
	at com.intellij.codeInspection.ex.GlobalInspectionContextImpl.lambda$runTools$12(GlobalInspectionContextImpl.java:338)
	at com.intellij.concurrency.JobLauncherImpl$1MyProcessQueueTask.lambda$call$0(JobLauncherImpl.java:297)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:608)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:683)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:639)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:607)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.concurrency.JobLauncherImpl$1MyProcessQueueTask.call(JobLauncherImpl.java:282)
	at com.intellij.concurrency.JobLauncherImpl.processQueue(JobLauncherImpl.java:350)
	at com.intellij.codeInspection.ex.GlobalInspectionContextImpl.runTools(GlobalInspectionContextImpl.java:394)
	at com.intellij.codeInspection.ex.GlobalInspectionContextBase.lambda$performInspectionsWithProgress$2(GlobalInspectionContextBase.java:273)
	at com.intellij.psi.impl.PsiManagerImpl.runInBatchFilesMode(PsiManagerImpl.java:451)
	at com.intellij.codeInspection.ex.GlobalInspectionContextBase.lambda$performInspectionsWithProgress$3(GlobalInspectionContextBase.java:267)
	at com.intellij.codeInspection.reference.RefManagerImpl.runInsideInspectionReadAction(RefManagerImpl.java:495)
	at com.intellij.codeInspection.ex.GlobalInspectionContextBase.performInspectionsWithProgress(GlobalInspectionContextBase.java:266)
	at com.intellij.codeInspection.ex.GlobalInspectionContextBase$2.run(GlobalInspectionContextBase.java:227)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:442)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsynchronously$5(CoreProgressManager.java:493)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:252)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:188)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$12(CoreProgressManager.java:608)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:683)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:639)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:607)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:175)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:252)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
```
- we provide `Unresolved reference` description although reference is not unresolved. Currently, it doesn't affect users because the corresponding message is not shown but anyway
- if `fixes` is empty, there is no reason to register anything at all

File: src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt [hunk 2]
```diff
@@ -62,6 +62,22 @@ class RsUnresolvedReferenceInspection : RsLocalInspectionTool() {
...
+            // enum E { A, B }
+            // match e {
+            //     A => {}
+            //     B => {}
+            // }   ^ RsPatBinding
+            override fun visitPatBinding(pat: RsPatBinding) {
+                if (pat.bindingMode != null) return
+                val patIdent = pat.parent as? RsPatIdent ?: return
+                val matchArm = patIdent.parent as? RsMatchArm ?: return
+                if (matchArm.pat != patIdent) return
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt:74]: What's about more complex patterns?
Like
```rust
mod inner {
    pub const C: i32 = 0;
}


fn main() {
    match (0, 0) {
        (C/*caret*/, _) => {}
    }
}
```

File: src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt [hunk 3]
```diff
@@ -62,6 +62,22 @@ class RsUnresolvedReferenceInspection : RsLocalInspectionTool() {
                 }
             }
 
+            // enum E { A, B }
+            // match e {
+            //     A => {}
+            //     B => {}
+            // }   ^ RsPatBinding
+            override fun visitPatBinding(pat: RsPatBinding) {
+                if (pat.bindingMode != null) return
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt:71]: Maybe add a test for this case and other corner cases as well?

**Review by Undin — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt [hunk 4]
```diff
@@ -62,6 +62,22 @@ class RsUnresolvedReferenceInspection : RsLocalInspectionTool() {
...
+                val patIdent = pat.parent as? RsPatIdent ?: return
+                val matchArm = patIdent.parent as? RsMatchArm ?: return
+                if (matchArm.pat != patIdent) return
+                if (pat.reference.multiResolve().isNotEmpty()) return
+
+                val context = AutoImportFix.findApplicableContext(pat)
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt:77]: It seems the corresponding code doesn't respect expected types because for the following code
```rust
mod inner {
    pub const C: i32 = 0;
}

enum E {
    C
}

fn main() {
    match 0 {
        C => {}
    }
}
```
enum variant is also suggested

image

File: src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt [hunk 5]
```diff
@@ -138,6 +138,12 @@ class AutoImportFix(element: RsElement, private val context: Context) :
             return Context(GENERAL_PATH, candidates)
         }
 
+        fun findApplicableContext(pat: RsPatBinding): Context? {
+            val importContext = ImportContext2.from(pat, ImportContext2.Type.AUTO_IMPORT) ?: return null
+            val candidates = ImportCandidatesCollector2.getImportCandidates(importContext, pat.referenceName)
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/inspections/import/AutoImportFix.kt:143]: Since we dropped old auto import code, can we remove `2` suffix now? Of course, not in this PR

File: src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt (hunk 4 above)
Comment (dima74) [reply src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt:77]: I propose to do it in another PR because it is not trivial (e.g. we can't just use `isCompatibleTypes` since it returns false for `&E` and `E`)

**Comment by dima74** [comment]
Moved the code to separate intention to not have "Unresolved reference" description and added more tests

File: src/test/kotlin/org/rust/ide/intentions/AddImportForMatchArmPatternIntentionTest.kt [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+    fun `test unavailable for field binding`() = doUnavailableTest("""
+        mod inner {
+            pub const C: i32 = 0;
+        }
+
+        struct Foo { C: i32 }
+        fn func(foo: Foo) {
+            match foo {
+                Foo { /*caret*/C } => {}
```
Comment (Undin) [inline src/test/kotlin/org/rust/ide/intentions/AddImportForMatchArmPatternIntentionTest.kt:101]: Actually, the intention can work in this case as well. Let's add a comment to explain current behaviour

File: src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt (hunk 4 above)
Comment (Undin) [reply src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt]: I'd add a test that verifies that intention provides several variants in this case and add a comment that ideally we need to filter out variants

Also, there is simpler example with similar problem:
```rust
mod inner {
    pub const C: i32 = 0;
}

struct Foo;


fn func(foo: Foo) {
    match foo {
        C => {}
    }
}
```

**Review by Undin — CHANGES_REQUESTED** [review]
I've just found that both `AddImportForMatchArmPatternIntention.invoke` and `AutoImportFix.invoke` are executed under write action. And it's wrong since it can show dialog to user that blocks any other read actions during user's choice

I suggest the following:
- add `override fun startInWriteAction(): Boolean = false` to this intention
- add `checkWriteAccessNotAllowed()` to `org.rust.ide.inspections.import.AutoImportFix#invoke(com.intellij.openapi.project.Project) and fix `AutoImportFix` in separate PR

File: src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt (hunk 4 above)
Comment (dima74) [reply src/main/kotlin/org/rust/ide/inspections/RsUnresolvedReferenceInspection.kt]: Added

File: src/main/kotlin/org/rust/ide/intentions/AddImportForMatchArmPatternIntention.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.isAncestorOf
+
+class AddImportForMatchArmPatternIntention : RsElementBaseIntentionAction<Context>(), PriorityAction {
+
+    override fun getText(): String = "Import"
+    override fun getFamilyName(): String = "Add import for path in match arm"
+
+    override fun getPriority(): Priority = Priority.TOP
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/intentions/AddImportForMatchArmPatternIntention.kt:28]: It seems it shouldn't have top priority.
It's ok for `Import` quick-fix for unresolved reference where users have to do something. But in this case, we intentionally don't show any annotation for this action so, I think, it reflects our uncertainty about what users want here

File: src/main/kotlin/org/rust/ide/intentions/AddImportForMatchArmPatternIntention.kt [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+import org.rust.lang.core.psi.RsPatBinding
+import org.rust.lang.core.psi.RsPatIdent
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.isAncestorOf
+
+class AddImportForMatchArmPatternIntention : RsElementBaseIntentionAction<Context>(), PriorityAction {
+
+    override fun getText(): String = "Import"
+    override fun getFamilyName(): String = "Add import for path in match arm"
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/intentions/AddImportForMatchArmPatternIntention.kt:26]: BTW why only match arms? There are other patterns as well where it can be useful
For example, `if let` construction

I don't force to do it in this PR but if you agree, let's change wording and user visible texts and `AddImportForMatchArmPatternIntention` class name

File: src/main/kotlin/org/rust/ide/intentions/AddImportForPatternIntention.kt [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import org.rust.lang.core.psi.RsPatBinding
+import org.rust.lang.core.psi.RsPatIdent
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.isAncestorOf
+
+class AddImportForPatternIntention : RsElementBaseIntentionAction<Context>() {
+
+    override fun getText(): String = "Import"
+    override fun getFamilyName(): String = "Add import for path in binding"
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/intentions/AddImportForPatternIntention.kt:24]:
```suggestion
    override fun getFamilyName(): String = "Add import for path in pattern"
```

File: src/test/kotlin/org/rust/ide/inspections/import/AutoImportFixTestBase.kt [hunk 10]
```diff
@@ -113,3 +96,23 @@ abstract class AutoImportFixTestBase : RsInspectionsTestBase(RsUnresolvedReferen
...
 }
+
+fun checkAutoImportWithMultipleChoice(expectedElements: List<String>, choice: String?, action: () -> Unit) {
+    var areElementEquals: Boolean? = null
+
+    withMockImportItemUi(object : ImportItemUi {
+        override fun chooseItem(items: List<ImportCandidate>, callback: (ImportCandidate) -> Unit) {
+            val actualItems = items.map { it.info.usePath }
+            areElementEquals = expectedElements == actualItems
+            BasePlatformTestCase.assertEquals(expectedElements, actualItems)  // exception here does not fail the test
```
Comment (Undin) [inline src/test/kotlin/org/rust/ide/inspections/import/AutoImportFixTestBase.kt:107]:
```suggestion
           assertEquals(expectedElements, actualItems)  // exception here does not fail the test
```

Don't forget to remove unnecessary import after this change

**Review by Undin — APPROVED** [review]

**Comment by dima74** [comment]
bors r=Undin

**Comment by dima74** [comment]
bors retry

**Comment by dima74** [comment]
bors retry
